### PR TITLE
Thumbcoil

### PR DESCRIPTION
**Pull request by imbcmdth**
Don't merge/publish yet!

### CODE REVIEW

File: source/_drafts/Introducing-Thumbcoil.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+In addition, the tools have the ability to show you the information contained within the most important NAL-units that make up the H.264 bitstream. Ever wonder what kind of secret information the video encoder has squirreled away for decoders to use? Now, with Thumbcoil, you can finally see for yourself!
+
+## Motivation
+{% pullquote right %}
+An H.264-encoded bitstream is composed of what are called **NAL units** or network-abstraction-layer. NALs are a simple packet format designed to use bits as efficiently as possible.
+{% endpullquote %}
+
+Believe it or not, there are very few good tools to generate a somewhat graphical display of the structure of media containers and the data that they contain. Debugging problems with video playback is usually a tedious task involving various esoteric *FFmpeg* and *FFprobe* incantations. Unfortunately at it's best, *FFprobe* is only able to print out a small portion of he data we were interested in.
```
Comment (forbesjo) [inline source/_drafts/Introducing-Thumbcoil.md]: > ... portion of **the** data we were interested in.

File: source/_drafts/Introducing-Thumbcoil.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+{% pullquote right %}
+A **transmuxer** takes media contained in some file format, extracts the raw compressed video and audio from inside (a process called demuxing) and repackages the compressed data  into another format (termed remuxing) without performing any re-compression.
+{% endpullquote %}
+
+## In the Beginning
+
+While building [Mux.js](https://github.com/videojs/mux.js) - the transmuxer at the heart of videojs-contrib-hls - we faced a problem: How do we determine if the output from Mux.js is correct?
+
+Early on we managed to figure out how to coax FFmpeg into creating fragmented-MP4s from TS segments that would play back in a browser with media-source-extensions (at the time that meant only Chrome.) But we needed a simple way to compare the output of our transmuxer with what was produced by FFmpeg. The comparison had to be aware of the MP4 format since the two outputs are extremely unlikely to be byte-identical.
```
Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md]: we potentially would want to define what `fragmented-MP4s` and `TS segments` are.

MSE isn't hyphenated. We probably should also link it to https://en.wikipedia.org/wiki/Media_Source_Extensions and/or the spec http://www.w3.org/TR/2014/CR-media-source-20140717/

File: source/_drafts/Introducing-Thumbcoil.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+{% pullquote left %}
+MP4 files are composed of **boxes** - hierarchical logical units that, conveniently, all start with a 32-bit length and a 32-bit box-type. Boxes will often contain other sub-boxes.
+{% endpullquote %}
+
+The answer to that problem was to build an "mp4-inspector" - a tool that would parse MP4 and display a sort of JSON-like dump of any relevant boxes and their contents. By generating a dump of the output from Mux.js and comparing it to a known-good fragment generated with FFmpeg, we could see where our transmuxer's output differed.
+
+The "mp4-inspector" was built as a web page so that we can have a graphical color-coded diff of the two segments. Over time the page gained an video element and we started appending the results of transmuxing segments directly into the video element's media-source to aid in instant feedback and validation of changes to Mux.js.
```
Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md]: `gained *a* video element`

File: source/_drafts/Introducing-Thumbcoil.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+## A Brave New World
+{% pullquote right %}
+A **media container** such as MP4 encapsulates the video and audio stream. It has metadata describing the streams, timing information for each frame, and the stream data itself.
+{% endpullquote %}
+
+As development continued, we would sometimes encounter streams that would fail in new and interesting ways. Some of these failures were, admittedly, due to bugs in Mux.js but as development continued and Mux.js itself became more robust they were increasingly caused by problems with the streams or issues with a particular implementation of the media-source-extensions specification.
```
Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md]: I don't think it's necessary to say that development continued in the second sentence. Just:

> Some of these failures were, admittedly, due to bugs in Mux.js; as Mux.js became more robut the failures were increasingly caused by problems with streams or issues with a particular MSE specification.

File: source/_drafts/Introducing-Thumbcoil.md (hunk 2 above)
Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md]: I would write out `Media Source Extensions (MSE)` and just use `MSE` in the rest of the article.

File: source/_drafts/Introducing-Thumbcoil.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+It eventually dawned on us that we really needed to learn more about what was happening inside of those videos. We needed to see not just what was happening at the media container level but we had to go deeper - we needed to peek into the video data itself. For that purpose we created Thumbcoil.
+
+{% pullquote left %}
+Inside of a container, video and audio are contained in data called **bitstreams**. Bitstreams are the data produced by encoders to represent the audio signals or video frames. Some common bitstreams are AAC for audio and H.264 for video.
+{% endpullquote %}
+
+## Thumbcoil
+
+Thumbcoil is a suite of tools designed to give you a peek into the internals of H.264 video bitstreams contained inside either an MPEG4 or MPEG2-TS container files. Using the tools in Thumbcoil you can get a detailed view of the internal structure of the two supported media container formats.
```
Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md]: Maybe like to the h264, mpeg4, and mpeg2-ts wikipedia pages?

Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md]: Also, above you called them just MP4s and TS segments but here you're calling them MPEG4 and MPEG2-TS container files.

File: source/_drafts/Introducing-Thumbcoil.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+In addition, the tools have the ability to show you the information contained within the most important NAL-units that make up the H.264 bitstream. Ever wonder what kind of secret information the video encoder has squirreled away for decoders to use? Now, with Thumbcoil, you can finally see for yourself!
```
Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md]: Should we reference the aside in cases like NAL-units?

> ... NAL-units (see aside) ...

File: source/_drafts/Introducing-Thumbcoil.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+## Motivation
+{% pullquote right %}
+An H.264-encoded bitstream is composed of what are called **NAL units** or network-abstraction-layer. NALs are a simple packet format designed to use bits as efficiently as possible.
```
Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md]: No need for hyphenation.
Also, "NAL units" would end up being "network abstraction layer units".

File: source/_drafts/Introducing-Thumbcoil.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+{% endpullquote %}
+
+Thumbcoil not only provides parameter set information in excruciating detail but also keeps the information with its surrounding context - the boxes it was contained by or the frame it was specified along with. This context is often very important to understanding issues or peculiarities in streams.
+
+## Built Upon Fancy Stuff
+
+One of the more interesting things about how Thumbcoil parses parameter sets is that is builds what is internally called a "codec" for each NAL unit type. These codecs are specified using what is essentially a fancy [parser combinator](https://en.wikipedia.org/wiki/Parser_combinator)-type setup.
+
+{% pullquote right %}
+Much of the data in the two parameter sets are stored using a method called **exponential-golomb encoding.** This method uses a variable number of bits to store numbers and is particularly suited to values that tends to be small.
```
Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md]: definitely link to the expbolomb wikipedia page.

File: source/_drafts/Introducing-Thumbcoil.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
+---
+title: Introducing Thumbcoil
+author:
+  name: Jon-Carlos Rivera
+  github: imbcmdth
+tags:
```
Comment (gkatsev) [inline source/_drafts/Introducing-Thumbcoil.md:6]: suggested tags:

``` yml
tags:
  - thumbcoil
  - mse
  - hls
```

**Comment by gkatsev** [comment]
You're missing a section, or at least links, about the website and the github page!

**Comment by imbcmdth** [comment]
I wonder if we can automatically create a links/biblio section on all posts from it's links with a plugin? Might be handy to have in all posts. But I will definitely add it.

File: source/_drafts/Introducing-Thumbcoil.md (hunk 1 above)
Comment (imbcmdth) [inline source/_drafts/Introducing-Thumbcoil.md]: He-Data is like the He-Man of information science.

File: source/_drafts/Introducing-Thumbcoil.md (hunk 9 above)
Comment (imbcmdth) [reply source/_drafts/Introducing-Thumbcoil.md:6]: Done.

**Comment by imbcmdth** [comment]
I added a final "Hey, it's also OSS." to the post with links to the repo.

**Pull request merged**